A Proposal to create a Sig Embedded (#79)
bytecodealliance/governance: SIGs/SIG-embedded/proposal.md

# Proposal to Create a SIG-Embedded and Industrial Applications

This document is a proposal to create SIG-Embedded and Industrial Applications as a formal Special Interest Group (SIG) under the auspices of the TSC of the Bytecode Alliance, as specified in the TSC’s charter. 

The Embedded ecosystem is ideally placed to benefit from the use of WASM and WASI. It has a set of constraints on performance in comparison to native non—WASM execution, determinism and reliability which can exceed those of other ecosystems. This gives rise to the definition of a low footprint device. The best way to address these specific needs is with a unified ecosystem of tools, technologies, documentation, and support. 

> The purpose of this group is to support and promote the use of WebAssembly and WASI in Embedded and Industrial user cases. This will be achieved by providing notes on best practice to existing practitioners, and by providing the TSC, WASI subgroup and WASM community group with uses cases which illustrate constraints and proposals to address such limitations.

Reviewer: ```suggestion
The purpose of this group is to support and promote security, correctness, modularity, and language compatibility in the Embedded and Industrial ecosystem through the use of WebAssembly, WASI, the component model, and WIT interfaces. This will be achieved by providing notes on best practices to existing practitioners, and by gathering use cases and constraints of the embedded space to guide new proposals to address existing limitations.
```

Reviewer 2: a couple of points, and I'm laying them out for suggestions how you would suggest we word this to be both clear and suggestive of what CAN be done. 

- the point at which it is simply not worth it to use components, let alone webassembly, in favor of artisanal code (my phrase for "we're rolling our own assembler here, baby") is one of the dividing lines between "we are figuring out how we should do this with components" and "components have little to no benefit in this space". I believe that the overall objective is to drive components and the evolution of the model such that it can expand its utility as small and optimized as it can go before the benefits vanish. This might be seen as a "negative" statement, but I believe the major point is to bring components as far down as we can whilst being explicit about where the "utility event horizon" is. 
- so I like the proposed wording change [here](https://github.com/bytecodealliance/governance/pull/79#discussion_r1516526243) in that it does say that, but I'd add the "utility event horizon" statement as well. Something like the last sentence satisfies that need to me.

If we just say "webassembly" and embedded/industrial, that's not driving components, which is the mission the BCA is here to support. So components -- which for me is shorthand for the security, correctness, modularity, and language compatibility of Nick's phrasing -- is front and center, knowing that there will always be some large swath of computing for which wasm is not appropriate. Write once, run 75% of the places. :-P

Reviewer 3: Thanks for the comments! @tschneidereit @squillace  

There is no any objection to components. I totally support the component, however, we still need to ensure the SIG has the flexibility to create the most appropriate solution for the real problems and requirements. There are quite some concerns about the side effects of component components in the embedded space. Let's drive the concerns in the SIG, maybe we can contribute to the component for the embedded. I am sure the component component will speak for itself.

Reviewer 2: yup, totes get you. "concerns" typically means that engineering tradeoffs need to be made given scenarios. That's precisely why I wanted us to document that "utility event horizon" beyond which artisanal code is likely to live. That provides several things. 
1. Clear reasoning about where value lies for certain scenarios and where it doesn't. This helps new adopters as well when they enter an arena for the first time.
2. A clear challenge for the component model to meet and exceed that utility event horizon. This is good because, hey, a good challenge is a good thing. :-)

Is that a fair way of rephrasing your point? Or do I miss something?

Reviewer 4: It appears that there are embedded use case that don't leverage WIT APIs (judging from last year WasmCon talks on the subject, for example) - would this mean the SIG can only discuss those as a mean to convert them to components?

Author: We’re concerned because no other SIG has been asked to restate the values of the Bytecode Alliance. It feels to us that TSC is concerned that we as the lead organization and proposer of the SIG are not aligned with the values.  You should note that we have been a member for 3 years. A rather expensive undertaking if we were not aligned for the get go!

If we must restate the values of the Bytecode Alliance, then we should restate all of these values and not cherry pick a subset. 

> Our mission is to provide state-of-the-art foundations to develop runtime environments and language toolchains where security, efficiency, and modularity can all coexist across a wide range of devices and architectures. We enable innovation in compilers, runtimes, and tooling, focusing on fine-grained sandboxing, capabilities-based security, modularity, and standards such as WebAssembly and WASI.

See above, this also includes “efficiency” and supporting a “wide range of devices and architectures”.

While “modularity, and language interoperability” are important, in the IoT space unless the solution is efficient and can run on the target device then these values are moot.

We recall that when WIT was discussed as the interface definition language it was accepted that this could result in a Component Model rendering and a wasi-preview-1 rendering (avoiding the Canonical ABI), both are valid.

We are completely aligned with and fans of the goals of the Component Model, but there are limitations on the technical implementation of Component Model which might prevent its use on small devices. We would love to contribute to an open discussion on this and how we can address it together, without the feeling like we are not aligned with the values of the Bytecode Alliance. 

There is a no definitive hard “event horizon” for the component model and our shared Bytecode Alliance values, for instance as the size of a device continues to shrink its ability to execute a wide variety of programming languages will also reduce – Go or JavaScript just doesn’t fit! - However an efficient implementation of the component is still desirable and offers advantages in terms of security and ecosystem. For a concrete  example consider an ARM-Cortex M4 running FreeRTOS where the desired footprint of a runtime maybe around 150kb.

Reviewer 2: no one can control humans! they'll of course talk about what they want to talk about. What the TSC wants to point out is that the BCA mission and its SIGS _are about moving components to their logical extension_. I am a board member of the BCA, so I believe in that mission. But I would be the first to say, also, that there will be scenarios that components are just not needed -- they add no value worth the modifications to their usage. 

But I'd also say, _that's correct for today and this year_. In five years, it may well be that component evolution means far, far more scenarios can be reached productively than we currently think about. So NO, I wouldn't say that the SIG "can't talk about" anything but components. :-) It's that our charter is to see how many places we can make them work -- before we add in other solutions. This is, again, precisely why I want one task of the group to be to describe the use cases and then match component features that enable or benefit these and where that list begins to get thin. That makes it clear that we can and I intend to have conversations about where components do not make sense. I assume others will share their own views and experiences.

This is my two cents, anyway.

Reviewer 2: Re: 

> We’re concerned because no other SIG has been asked to restate the values of the Bytecode Alliance. It feels to us that TSC is concerned that we as the lead organization and proposer of the SIG are not aligned with the values. You should note that we have been a member for 3 years. A rather expensive undertaking if we were not aligned for the get go!

You are absolutely correct here. I'd urge the TSC to accept your suggestion for the wording. That is right and good.

Author: > no one can control humans! they'll of course talk about what they want to talk about. What the TSC wants to point out is that the BCA mission and its SIGS _are about moving components to their logical extension_. I am a board member of the BCA, so I believe in that mission. But I would be the first to say, also, that there will be scenarios that components are just not needed -- they add no value worth the modifications to their usage.
> 
> But I'd also say, _that's correct for today and this year_. In five years, it may well be that component evolution means far, far more scenarios can be reached productively than we currently think about. So NO, I wouldn't say that the SIG "can't talk about" anything but components. :-) It's that our charter is to see how many places we can make them work -- before we add in other solutions. This is, again, precisely why I want one task of the group to be to describe the use cases and then match component features that enable or benefit these and where that list begins to get thin. That makes it clear that we can and I intend to have conversations about where components do not make sense. I assume others will share their own views and experiences.
> 
> This is my two cents, anyway.

100% on board with moving components to their logical extension, I do think, that this could be possible to push their efficiency further within the embedded / industrial space. I'm also aware that we have folks who have already deployed wasi-preview-1 in embedded .and industrial devices, they'll need some support, and they'll also need an "on-ramp" that helps them transition to component model based solution. Supporting these folks is also important, it helps further the component model's adoption.

I understand the values, and get them, it's implicit in our participation. The concern is the over stipulation of a subset of them that prevents the SIG from looking at the technical areas we may address in order to further the cause of the component. 

> no one can control humans! they'll of course talk about what they want to talk about.
100%. I agree. I'm just keen to make sure that, the conversation, energy, and innovation that derives from talking is done within the BCA and helps to drive the component concept. It would be a real shame if the over stipulation made people feel that this isn't the place for their energy.

Author: > hey @woodsmc , this is your utility event horizon:
> 
> > There is a no definitive hard “event horizon” for the component model and our shared Bytecode Alliance values, for instance as the size of a device continues to shrink its ability to execute a wide variety of programming languages will also reduce – Go or JavaScript just doesn’t fit! - However an efficient implementation of the component is still desirable and offers advantages in terms of security and ecosystem. For a concrete example consider an ARM-Cortex M4 running FreeRTOS where the desired footprint of a runtime maybe around 150kb.
> 
> I have no concerns about "definitive" -- that's a red herring here. You just laid out one of the scenarios that may be beyond the ulility horizon of components. Good enough for me. :-) And I think that's fine and real and no problem. Bringing these up helps people understand tradeoffs in engineering. I WANT to have these _rough limits_ described as part of this work, which will necessarily include discussions of those things that are beyond those limits.

@squillace - 
There is a delta between where WebAssembly is useful and where the Component concept is useful. As device capability degrades various benefits of the component concept gradually start to become impossible to maintain, the primary one will be dictated by the size of the hosted Wasm module and it's resource requirements, and this implies the originating programming language. Even when the component concept is no longer viable, WebAssembly continues to deliver value, until, at last we reach a lower bar with Wasm2C, or perhaps eventually until some future implementation of a "Wasm2Arm32"... and this, I'm guessing is the gradual decline to "artisanal" code you describe?

It would be awesome to work on shirking this delta, so we can get the component model concept continuing to deliver value right down towards the smallest of devices. That's very exciting.

Reviewer 5: Hi everyone,
I just have to jump into the conversation at this point.

> There is a delta between where WebAssembly is useful and where the Component concept is useful
I highly agree to this.

>  "we're rolling our own assembler here, baby"
And this.

Especially in industrial applications using highly constrained devices, one could assume, that runtimes/compilers are are highly fine tuned to fit the current hardware and requirements. Depending on the use-case, where only host-module communication is involved and low latency is required, one could assume, that this may be one of the examples, where Components are not practical. Still, I see a huge benefit of describing the available APIs in WIT.

So I think we could still use the Component Model for this proposal. But when taking design decisions we must keep in mind that "workaround" solutions with a Non-Component Model adaptions (and custom modifications to the runtime/compiler) are probably a lot more common than in other areas, just because of the nature of highly constrained devices.

Reviewer 6: (Pls bear with me as I am not an english native speaker)
I am a _big_ fan of the component model as I know the struggle of „composing software artefacts“. I also experienced the pain of „managing“ software on resource constrained devices. I assume we all know that as computer scientists we have to be aware of the tradeoffs when striving for performance, low footprint, sandboxing and being polyglot. Sandboxing and „polyglot“ also mean reusability and manageability of software - sandboxing on its own will have benefits in our daily work. Now combine this „polyglot“ - even more beneficial. Bringing this to resource constraint devices blows my mind as these are the targets of the „embedded world“ - in the „embedded world“ the above tradeoffs hit really hard!
With the component model, „sticking software together“ is another boost to lower the efforts in software crafting. The _combination_ of all these properties imho is _so unique_!

When looking at the state of the art for managing software, docker comes into my mind. But it has tradeoffs - e.g. binding to Linux environments, which is not usable on very resource constraint devices. So, despite the fact that docker is cool, it is not usable for such devices -> „all arguments for using docker then fall apart“. And the _very same risk_ holds if the component model has technical implications and assumptions that _might_ turn out to be not feasible for resource constraint devices. It would be a pity, if this leads to a situation, where „WebAssembly as a technology (= a smart combination of compiler, byte code and runtimes)“ cannot be used for constraint devices, because - as depicted above - _WebAssembly can deliver huge benefits even w/o a fully fledged component model_. So, imho forcing/obligating the usage of a fully fledged component model, which _might_ not be feasible for the embedded world, feels to be wrong.

To me, insight the „eSIG“, we should _also_ talk about approaches of using WIT, but maybe find another way than the canonical ABI (Chris already mentioned the discussion about „based on canonical ABI“ and „based on wasi-preview1“). And we should also drive the discussion based on data, i.e. having a benchmark suite for regression tests. As of now, I am not aware of data regarding the impacts on performance and footprint when using the component model with the canonical ABI. If it turns out via evidence/proofs, that components are not doable (or not reasonable) on resource constrained devices then there should be easy ways to use alternative approaches - and the eSIG as part of the BA could think about this in tight collaboration with champions of the component model. And, as a third thing that comes directly into my mind: For the embedded world, imho, it is very important to also have maturity and longevity - such devices run for decades and more and more must be „managed“ (=sandbox is very important). And, of course, this list is not exhaustive…

Reviewer 7: @ttrenner I fully support your point here.

We should in this early state not jump to conclusions and setting an early bias by choosing a solution to our problem already.

Although things like the component model offer a lot of prospect and might even be the final solution after some tweaks, we should all keep an open mind at this stage.

So let's use this PR to talk about the start of the eSIG not the outcome.

Author: The original statement attempted two things:

1.	Define the purpose and hence rough scope of the SIG (first sentence)
2.	Define in broad strokes how the output of the SIG would support its purpose (second sentence).

The changes proposed by @fitzgen and as explained by @tschneidereit would appear to explicitly seek to define the solution space to the Component Model and WIT. I’m speculating that there may be some concern that without this stipulation that the SIG’s work could cause fragmentation within WASI. We tried to explicitly address this with text *“unified ecosystem of tools, technologies, documentation, and support”* which appears earlier in the proposal. That said, let’s look at the implication of changes suggested in some more detail.

I don’t think there is any objection to using WIT at all. The worry is around the imposition of the Component Model as the *only* way the SIG would address the embedded space. There are two issues:

1.	Existing installed based of wasi-preview-1; we need to continue to support these folks and provide them with a smooth transition to a future WASI standard. Not everyone can or will move from p-1, to p-2 and V1. Some will need to continue to use p-1 and will want to migrate to V1 once it’s ready. Everyone will move at their own speed. Here we can support with tooling, and interface definitions that would make the transition to “wasi proper” easy and smooth. After all, there is a much longer refresh rate in IoT than in cloud / hosted computing. We have seen Bytecode Alliance members talk of updates every 4 years. Here the SIG may want to step up and provide this support.

2.	Based on the current implementation of the Component Model there is concern that it may not be possible to scale this down to address the embedded market. As the discussion with @squillace points out as the capabilities of the embedded devices decrease there is a delta between when the component model will stop delivering value, and when WebAssembly will stop delivering value. The SIG should 100% look to use the component model, and this concern should be addressed. @ttrenner suggested a data driven approach where we evaluate the concern. The original second sentence sought to address this data driven approach by driving the SIG to preform investigations to support the embedded use cases and then propose solutions.  This approach is identical to what @xwang98 suggested, we’d like to work with address these concerns, and try to get this delta as small as possible, essentially, get the component model on as many embedded devices as possible. We should note that the entire industry is being aided by Moores-Law, devices are getting more powerful, so even for use cases today where it may not be possible to use component model, we can deliver value and a hook to encourage folks to migrate to the component model when feasible.

I fear that the implications of the proposed change would mean only looking at the Component Model – aka preview2 and beyond, not being able to work to support existing preview-1 users. The only full runtime we can use with preview2 is WasmTime and it’s currently not designed to target embedded systems, so we’d limit initial interest in the SIG, as there wouldn’t effectively be any Embedded system to speak of until that is addressed. An incremental approach is preferred. 

Just for clarity, there is no intention to reject the Component Model, in fact it’s the opposite. The desire to ensure we can use the component model on some of the most aggressively constrained devices we can target and bringing the ecosystem with us on this journey. Ultimately, this should help the entire Bytecode Alliance and the wider ecosystem. If this is acceptable then perhaps we can rephrase these two sentences to capture this ?

Reviewer 8: > As a user of wasi-preview1 on embedded systems, we agree with @woodsmc's opinion and hope that the specification can smoothly transition. We hope that the current device can continue to run wasm programs that meet the specifications, which is crucial to our product ecosystem.

Agree with @woodsmc too, millions of our products running with wasi-preview1 on very resource constrained platfrom (only ~48KB RAM available).

Reviewer 9: Thank you for all the replies here, and the excellent discussion and clarifications!

I'll try to address all the key themes, but will do so by replying to only some comments. This is still getting pretty long, so please bear with me 😄. And also, please read my replies as coming from me as a member of the TSC, but not as a fully agreed-upon TSC position, as we haven't had a chance to discuss all the points brought up here. And I'll start with the most fundamental aspect of why we're asking for clarifications here:

> We’re concerned because no other SIG has been asked to restate the values of the Bytecode Alliance. It feels to us that TSC is concerned that we as the lead organization and proposer of the SIG are not aligned with the values.

@woodsmc if we didn't trust you as the SIG's proposer, we'd not start addressing that by asking for more explicit language around the SIG's goals in the proposal. Please do read my statements as written, in particular where I say `What we want to see made explicit in the SIG's charter is something we of course all know [..]` and `we wanted to make that more explicit to help people new to this space to more readily understand things.` I in particular fully agree with what you're saying in the last paragraph of [this reply](#discussion_r1518023821), and appreciate your approach here.

One thing I want to express here for the benefit of the many interested parties who're not as familiar with the Bytecode Alliance is that the BA's TSC is not meant to be a purely facilitating infrastructural body: it's meant to provide overarching technical guidance to hosted activities, and to provide clarity in this regard. It is with that in mind that we're providing this input. And the embedded space has some specific aspects to it that make it particularly important to provide clear guidance. I want to emphasize that based on many previous conversations we understand that you, Chris, understand the goals of our work very well. As I emphasized, we want to ensure that the SIG's charter provides clarity for participants coming from very different directions.

Since we provided this feedback last week, the discussion here IMO has highlighted that indeed we do need to have a more thorough discussion about how best to frame the SIG's scope and the TSC's guidance. More on that below.

> There is a delta between where WebAssembly is useful and where the Component concept is useful

This is probably the most important aspect to address: I don't think that there's a fundamental reason why the Component Model in its approach needs to be more heavyweight than core WebAssembly. I'm aware that in its current iteration, that is the case, but there are architectural and implementation improvements to be had that can address most, and *maybe* all of them.

That will certainly not entail bringing the full generality of the component model to even the smallest of embedded use cases, same as some of these use cases will end up subsetting core wasm itself. But it should entail integrating this subsetting into the architecture in as seamless a manner as possible, so that unified toolchains can be used, and so that we achieve as much portability between different environments as possible.

I actually like the part where the proposal currently talks about `providing the TSC, WASI subgroup and WASM community group with uses cases which illustrate constraints and proposals to address such limitations` and think we should keep it in, because I see this as a pretty key part of what this group should do.

> It appears that there are embedded use case that don't leverage WIT APIs (judging from last year WasmCon talks on the subject, for example) - would this mean the SIG can only discuss those as a mean to convert them to components?

and

> I'm also aware that we have folks who have already deployed wasi-preview-1 in embedded .and industrial devices, they'll need some support, and they'll also need an "on-ramp" that helps them transition to component model based solution. Supporting these folks is also important, it helps further the component model's adoption.

These are excellent questions and points! My (not the TSC's agreed-upon) tentative answer would to say that of course the SIG can discuss these use cases, but should as one of its key activities try to come up with ways to make these use cases work with general-purpose, WIT based toolchains as much as possible—and as currently expressed in the proposal provide input into the standards process where it identifies constraints that make that infeasible. And providing on-ramps to component tooling specifically is I think 100% in alignment with our expectations.

> [..] there will be scenarios that components are just not needed -- they add no value worth the modifications to their usage.

@squillace, I think it's more subtle than that. As emphasized above, I agree that in their current iteration and existing implementations, there are limits to how far "down" (into the embedded space) components can go. I agree with Chris' optimism about pushing those limits down a lot further, and wouldn't rule out that we can push them as far down as it'll make sense to push WebAssembly itself.

I also think that there is another source of value which is hugely important, and which I alluded to earlier: the unification of toolchains. The Bytecode Alliance set out to build new foundations for software development—and we don't want these foundations to be insular wherever it can be avoided. To us that means enabling developers to use a single toolchain to target as many different environments as possible. And ideally in a way where at least parts of their applications can be directly reused in these different environments by composing them in different ways.

This all to me means that even if an environment cannot make use of all the capabilities of components—for example because it's too small to ship a GC, and hence can't support components targeting Wasm GC instead of linear memory—, it should still be possible to target that environment with a toolchain that produces artifacts that run in more fully-featured environments. And I think the best way to get there is by targeting the concept of components, even if for some specific environments we ultimately need to shave off many of its capabilities.

> I'm just keen to make sure that, the conversation, energy, and innovation that derives from talking is done within the BCA and helps to drive the component concept. It would be a real shame if the over stipulation made people feel that this isn't the place for their energy.

I agree! And I'm confident that through these discussions we'll arrive at a scope for the SIG that supports this very well, and provides newcomers with the desired clarity of the aims here.

> I fear that the implications of the proposed change would mean only looking at the Component Model – aka preview2 and beyond, not being able to work to support existing preview-1 users. The only full runtime we can use with preview2 is WasmTime and it’s currently not designed to target embedded systems, so we’d limit initial interest in the SIG, as there wouldn’t effectively be any Embedded system to speak of until that is addressed. An incremental approach is preferred.

We very much understand that there is an existing base of real-world usage of WebAssembly built on WASI Preview1, as well as entirely proprietary approaches. What we (again: in my informed interpretation of the TSC's consensus) want to avoid is that the SIG becomes primarily focused on incremental work building on these improvements. We fundamentally believe that the future is substantially bigger than the past when it comes to WebAssembly's use everywhere, including in the embedded space, and we want to encourage the SIG to focus the majority of its efforts on helping to build that future.

I've written half a book here, so I'll stop now, hoping that I hit on all the major themes in the discussion.

I do think we should further tweak the wording in the proposal, and ensure that we incorporate the input here.

Reviewer: Our [operational principles] &mdash; a document coauthored by you, @xwang98, in your time on the proto-TSC &mdash; makes it crystal clear that in the Bytecode Alliance we collaborate with the relevant standards bodies to implement interoperable, standards-compliant software and that "a key goal for us is a reduction of fragmentation across the ecosystem".

The relevant standards body for Wasm APIs that are not Web-specific, such as those for the embedded space, is the Wasm CG's WASI subgroup. The WASI subgroup standardizes APIs defined by WIT interfaces and therefore, because WIT is an IDL for the component model, these APIs *must* be based on the component model.

That said, I do not believe this constraint is onerous. A component that contains code which operates, for example, within a single linear memory should not impose any additional overhead compared to an equivalent core module that contains code which also operates within a single linear memory. That statement is prescriptive, but I already believe it to be true in nearly all cases right now. My understanding of this proposed SIG was that its responsibilities would include identifying, via the concrete use cases of the embedded space, any scenarios where standards, runtimes, or toolchains do not live up to that statement. And, having identified such a scenario, that the SIG would collaborate with involved stakeholders (whether standards editors, runtime maintainers, or toolchain developers) to design and implement fixes that resolve those discrepancies.

In the past we've discussed how to best integrate `libc`'s `read` method with WASI's I/O interfaces such that we can ensure that the user's pre-allocated buffer passed into `read` is sufficient, and that the component model and its canonical ABI do not force additional allocations. Other performance-related discussions have questioned whether it is possible for Wasm guests to use efficient bump allocation for temporary data received over the canonical ABI, should they choose to. My understanding is that, in both cases, we concluded that the existing semantics of the component model and its canonical ABI already suffice: `read` can use its pre-allocated buffer and Wasm guests have the flexibility to bump allocate. Had those discussions instead concluded that, for example, we cannot implement `read` without additional allocations, then tolerating that conclusion would not have been acceptable. The correct and acceptable response would have been to diagnose why `read` must allocate, determine what changes need to be made (whether in standards, runtimes, or toolchains) such that we can avoid it, and finally implement those changes. If we find flaws or limitations in the component model, let's work together to fix them.

Collaborating to identify and fix flaws in our shared foundations benefits everyone in our ecosystem. Ensuring that the component model satisfies the particular performance requirements of the embedded space also benefits people developing higher up the stack, when their apps allocate a little less and are a little more responsive than they otherwise might be. And maybe those developers that are higher up the stack are laser-focused on developer experience, polishing and removing paper cuts from component model toolchains. Embedded developers also benefit from that toolchain polish. Now, because performance and toolchains are attractive, even more developers join our ecosystem, and they start contributing in their own unique ways, bringing additional benefits to all existing community members. This is how we create a virtuous cycle. A rising tide that lifts all boats.

In contrast, forking our shared foundations, introducing competing standards, and shipping proprietary, non-standard APIs all increase ecosystem fragmentation. These things are in direct conflict with the Bytecode Alliance's operational principles and mission.

I must admit that I am confused and surprised by the push back in this thread. In fact, there have already been discussions among the proposed SIG's members about a possible new core Wasm proposal to allow for custom, non-64KiB linear memory page sizes. This is an instance of exactly that identify-and-resolve-issues approach I described above, although applied to the core Wasm language rather than the component model. The framing of that discussion was *not* "Wasm is unacceptable for embedded because of its 64KiB page limit, so therefore let's build something that is not Wasm and does not have the 64KiB page size limit but is otherwise similar." That zero-sum, all-or-nothing, not-invented-here approach would fall flat on its face. Instead, the framing of that discussion correctly was "let's work with existing stakeholders to improve Wasm and resolve this 64KiB page size issue we've identified." This latter framing leads to solutions that improve our shared commons, benefiting everyone. So why would we take that positive-sum approach for the core Wasm language but not for the component model?

It is incredibly defeatist to give up and throw out the whole shared foundation that is the component model, fragmenting the non-Web-specific Wasm ecosystem, before we have attempted to resolve &mdash; let alone even benchmarked, identified, or analyzed! &mdash; any supposed overheads it imposes. I cannot endorse any effort launched from such flawed origins.

[operational principles]: https://github.com/bytecodealliance/governance/blob/main/operational-principles.md

Reviewer 10: Given the many excellent points made on this PR, I'll add a non-technical reason for my concern surrounding the TSC's proposed change.

In context, the addition of the component model wording is worrisome for those of us working on deeply embedded systems. Much of our work involves interfacing Wasm with domain specific middlewares and has not yet moved to the component model. In contrast, I have been in WASI meetings where topics that did not relate to the component model were tabled. I get it-- folks are trying to push a WASI standard forward and there is only so much time in these meetings. However, I think the draw of the embedded SIG is the opportunity to look closer to the hardware and to tightly incorporate the component model only if it fits a given use case.

At the end of the day, best practices that minimize the elbow grease to run Wasm modules on both embedded platforms and fully fledged environments will likely involve aligning with the component model. However, it should be explicit in the eSIG charter that there is room for non-component discussion that still points towards the shared goals of security, modularity, etc.

Reviewer 11: @woodsmc

> [...] After all, there is a much longer refresh rate in IoT than in cloud / hosted computing. We have seen Bytecode Alliance members talk of updates every 4 years. Here the SIG may want to step up and provide this support.

@no1wudi

> Agree with @woodsmc too, millions of our products running with wasi-preview1 on very resource constrained platform (only ~48KB RAM available).

I also agree. What if, in addition to @fitzgen's proposed change here, we also add to the charter "We will support existing WASIp1 interfaces and tools, including WASIp1 support in wasi-sdk and wasi-libc, for 4 years"?

That way, we make it clear that we're committed to supporting existing users.

Develop new interfaces and new tools is a separate question. Here, we don't yet have 4+ year upgrade cycles to consider, and we can consider questions of fragmentation and what we want to support for new 4+ year upgrade cycles. I also agree with incremental approaches:

> An incremental approach is preferred.

There is an incremental path that we discussed in WASI; the idea is to use the Canonical ABI rendering of a Wit API to obtain a P1-style API. For an example of what this looks like, in my [hello-embedded](https://github.com/sunfishcode/hello-embedded) repo, Run `wit-bindgen c wit --world=blink` and look for function declarations marked with `__attribute__((__import_module__("..."), __import_name__("...")))` and `__attribute__((__export_name__("...")))` in the generated blink.c file. These describe a core-Wasm C ABI, that could be directly implemented by engines with just P1-style features. There are various dimensions along which the code could be improved that we could work on.

So the incremental path we discussed in WASI is, someone could build a similar tool which outputs witx descriptions of these APIs, which would support an incremental implementation approach.

Reviewer 9: > At the end of the day, best practices that minimize the elbow grease to run Wasm modules on both embedded platforms and fully fledged environments will likely involve aligning with the component model. However, it should be explicit in the eSIG charter that there is room for non-component discussion that still points towards the shared goals of security, modularity, etc.

@kangaruppel I think this gets to the heart of what we in the TSC are trying to emphasize, as expressed above by Nick and me: we are explicitly asking this SIG to not go into things with preconceived notions on where the component model is or isn't a good fit. And based on previous conversations around Wasm usage in the embedded space, we are concerned that there would be a default towards this that'd go beyond what we'd be comfortable with. That doesn't mean we're not leaving any room for other discussions, but we we *do* want to ask that that room only be used once it really is clear that the constraints (of whatever exact nature) rule out a path towards components, instead of prematurely segmenting the design space into "components as they are now work or don't work".

In all of this, we certainly understand real-world constraints, and that they're not always technical in nature, and thus all we're asking for is a heavy bias towards forward-looking interoperability, and emphatically *not* an absolutist take.

As I mentioned before, I agree that we should further tweak the language in the proposal to properly capture this, and that our current proposed change isn't the final word.

Author: @fitzgen I totally agree with you on avoiding fragmentation, and taking a pragmatic approach for improvement. That was the original proposals intent. No one wants to see complete fragmentation - the risks of dissipation are too high. You are right to oppose this. We should note that there is controlled fragmentation, like the GC / non-CG profiles etc. But this as pointed out, can be achieved with a unified toolset, which is awesome.

WASI has already been hugely successful in the embedded space! - The BCA should be amazingly proud of this achievement - "millions of devices" as @no1wudi points out, and that was just from one participant, we already know that there are other organizations using preview 1 in production. That's something that should really be acknowledged and celebrated - its a huge impact. And needless to say a huge financial commitment to WASI that's been made by these organizations. What a vote in confidence that is! – The whole of the BCA should be very proud indeed, as software professionals the opportunities we get to impact so may people and so many devices is so very rare.

The challenge now is that we've huge installed base of preview1 running on tightly constrained devices, so how do we now help them adopt the Component Model ? - As you rightly point out we should take a pragmatic evidence driven approach to do this, and this is what was originally proposed - or at least the intention of the wording. I recognize that sometimes the intention and words written do not always align, and these discussions are great to sorting that out.

The issue, IMHO, is that the proposed changes and the accompanying comment imply that the Component Model is the *only thing* the SIG could look at, and to question the Component Model in anyway - even it's suitability for the embedded space in it's current incarnation, would somehow be heretical and against the values of Bytecode Alliance. 

Yikes! – Imagine the implications of this - we spend 2 years, maybe longer looking at performance improvements to the component model, limited by the constraints of it's current implementation and binary interface . Canonical ABI. Meanwhile what happens to the installed base of p1? - They've got a compiler tool chain which will eventually be deprecated and a format which is moving on faster than they can keep up…. And goodness knows how much money and corporate risk tied up. Will they wait? Or Will they go off and try to address this themselves? – We’d end up with the fragmentation we all want to avoid.

I’m not sure that was what was intended by the proposed change and comment, but it is what is being heard. That explains, in my opinion why this important conversation is so lively.

Author: @tschneidereit 

>emphatically not an absolutist take.

Awesome, thank you for stating this.

Author: @sunfishcode  - Dan, they are some awesome ideas. 

>I also agree. What if, in addition to @fitzgen's proposed change here, we also add to the charter "We will support existing WASIp1 interfaces and tools, including WASIp1 support in wasi-sdk and wasi-libc, for 4 years"?

I had a similar thought, then second guessed myself, as perhaps rather than a time frame a conditional trigger would be better, etc? - And trying to address the needs to all of the Embedded users is tough.

It made me think we'd need more time to come up with a great transition and incremental roadmap. Maybe, we could add that the SIG should have a role in helping to define and implement that ? - This way, we give more time, and space to get feedback from the wider embedded stakeholders?

I mention implement, as I can imagine the roadmap may also see a migration of maintenance of various tools between contributing partners. 

What do you think?

Reviewer 9: > The issue, IMHO, is that the proposed changes and the accompanying comment imply that the Component Model is the _only thing_ the SIG could look at, and to question the Component Model in anyway - even it's suitability for the embedded space in it's current incarnation, would somehow be heretical and against the values of Bytecode Alliance.

I think it makes most sense to from hereon out lead this conversation with the assumption that all sides are open to changing the language, and maybe even to not focus on either the preexisting version or the proposed one as a starting point.

I think it also makes most sense to see the situation as one where there might be disagreements on some aspects (to wit, I'm not convinced that there's 100% within the TSC on every detail), but where we're all aligned on the points that truly matter: we want WebAssembly to succeed in the embedded space, and we want it to succeed in a way that truly improves upon some key problems around security and interoperability. As @woodsmc says, we've already achieved a lot here, and we have amazing foundations for achieving much, much more.

I also think that we're getting closer to a shared understanding here, which is great!

I think one source of what's perhaps a bit of a disconnect is that on the standardization side we always saw—and talked about—WASI Preview1 as an interim step towards a stable standard, instead of a long-term supported platform. At the same time, people were (understandably!) eager to employ what was there for industry use cases. Working on eliminating that disconnect and ultimately getting the embedded space onto a stable platform seems like it could be a key aspect of eSIG's agenda.

One thing I realize I should strongly emphasize: **nobody on the TSC is claiming to know better about the embedded space than the actual practitioners in this space!** Y'all are the ultimate experts on what the constraints are, and we don't intend to question your word on what is and isn't possible—that's not what any of our feedback is about!

I'll try to make very explicit some key aspects of my understanding of things:
1. The TSC is asking eSIG participants to explore how far the component model can be taken in the embedded space
2. As made explicit in the current wording in the proposal, we're asking eSIG to report back to the standardization process on use cases that aren't currently covered and ideally (but *not* necessarily) proposals for how to address them
3. I'm 100% certain that the current iteration of the component model will *not* cover many important use cases, and in no way think eSIG should operate on some kind of assumption that the component model is immutable
4. We also understand that industrial use cases need stability and clear roadmaps, and I think it makes a lot of sense for eSIG to work on these aspects, e.g. along the lines of what @sunfishcode described

If that makes sense to folks, maybe we can next try to incorporate key outcomes of this discussion into the proposal?

Reviewer 2: at this point, I keep reading eSIG as a kind of vape cigarette

Reviewer: > millions of our products running with wasi-preview1

> The challenge now is that we've huge installed base of preview1 running on tightly constrained devices, so how do we now help them adopt the Component Model ?

To be clear, everything I've been saying, I've been saying in the context of *new* APIs, which by definition are not available on existing deployments. To summarize:

* New APIs must be standardized, per BA operational principles and values.

* The relevant standards body is the WASI subgroup.

* WASI APIs are defined in WIT.

* WIT is an IDL for the component model and its canonical ABI.

* Therefore, new APIs must be based on the component model and its canonical ABI.

* If that leads to technical problems, then let's address those technical problems together! As Till said, we should not assume that the component model and its canonical ABI are immutable. We can, should, and will design and implement fixes for issues that come up. We *actively want* your domain expertise to help shape the solutions here, and are simply asking you to participate in the standards process so that everyone, including yourselves, benefits.

As for preview 1 (which was, as Till noted, never intended to be nor advertised as a long-term supported platform) and existing preview 1 deployments: no one is saying that people can't maintain legacy toolchains, fix bugs, and improve performance, nor that the SIG itself cannot discuss and organize around those things.

*Just that new APIs must go through the WASI standardization process*.

> Meanwhile what happens to the installed base of p1?

And as far as migrating that existing p1 install base to p2 so that they can take advantage of new APIs, Dan [sketched an incremental migration path in this comment](https://github.com/bytecodealliance/governance/pull/79#discussion_r1519974402). Use the raw canonical ABI version of the new WIT APIs from core Wasm; extend the legacy toolchain to support this functionality; finally, when runtimes can be updated, make the full migration. I'd love it if one of the SIG's responsibilities, formally outlined in this proposal, was to flesh this plan out and lead the way on implementation. Or do that for another viable, incremental migration plan.

-----------------------------------

> not focus on either the preexisting version or the proposed [version of this
> document] as a starting point.
>
> [...]
>
> maybe we can next try to incorporate key outcomes of this discussion into the proposal?

Big +1 from me.

Reviewer 12: > I also agree. What if, in addition to @fitzgen's proposed change here, we also add to the charter "We will support existing WASIp1 interfaces and tools, including WASIp1 support in wasi-sdk and wasi-libc, for 4 years"?
>
> That way, we make it clear that we're committed to supporting existing users.

I'd suggest avoiding here implementation details and rather define what problems we solve. The problem is that there's a number of existing installations of runtimes with preview1 support that can't be updated, and we need to find a way to support them.
Supporting WASIp1 in wasi-libc is a solution, one of a few (for this specific  one, I'm drafting a proposal for the p2 canonical ABI to p1 adapter). I'm all for solving this particular problem (in fact, my team has exactly that problem to solve), but I don't want to limit our options for ways of solving the problem. I think we should just mention that we will attempt solve this problem (somehow).

Author: @fitzgen @tschneidereit Totally happy to revise the text, and to think openly about what this should include. Thank you explicitly calling that out.

@fitzgen
> * Therefore, new APIs must be based on the component model and its canonical ABI.

On this point, we’ve a very different take. I want to air it. It’s the use of the word *“must”*, I don’t think that’s quite right as this doesn’t reflect what was previously agreed. What was agreed was that WIT could be used to generate WITx and this would generate preview1 bindings. We recall that conversation. Intel raised the point and it was agreed. I know that while agreed, the tooling for this hasn’t been built. - It is important that this continue to be a possibility. This provides an important part in the transition from P1 to new iterations for WASI. That line should instead read:

* Therefore, new APIs will be based on the component model and its canonical ABI, and can, optionally also be rendered in Witx, and hence suitable for p1 interfaces.

@tschneidereit
>That doesn't mean we're not leaving any room for other discussions, but we we do want to ask that that room only be used once it really is clear that the constraints (of whatever exact nature) rule out a path towards components, instead of prematurely segmenting the design space into "components as they are now work or don't work".

As technical guidance the TSC can of course suggest that this approach be taken, but mandating it is a problem. I can foresee disagreements around what the constraints are, and if enough of an investigation had been conducted. Likewise, what would happen if there was no contributor who offers to undertake the task? Or what if that contributor undertakes the task at a very slow pace? – The prioritization would prevent other work from proceeding, like for instance the continued support of new interfaces on the existing installed base. 

A set of parallel, related, but decoupled tracks would be better. One aimed at providing source compatibility for new interfaces using the previously agreed wit to witx approach, and a second aimed at validating / tuning the component model on embedded platforms. This way we continue to provide for existing p1 systems and allow the Component Model the space and time to find it’s spot in the embedded landscape. 

The source compatibility means that we’d have a natural and easy transition to the component model for those use cases where the advantages of the technology make sense. It also means that P1 only users can still participate in standards discussions with interfaces they can actively use and evaluate. Without this they would effectively be locked out of API discussions via a lack of accessible technology.

This approach avoids any forgone conclusions as to the role of the component model, its suitability or performance characteristics. Likewise, it doesn’t impose or mandate it’s use without the data points which validate its suitability. 

@fitzgen 

> As for preview 1 (which was, as Till noted, never intended to be nor advertised as a long-term supported platform) and existing preview 1 deployments: no one is saying that people can't maintain legacy toolchains, fix bugs, and improve performance, nor that the SIG itself cannot discuss and organize around those things.

There is a technical debt around P1, and as evident in the discussions here you can see future SIG members acknowledge that and start to look for solutions. But the same costs will be born on P2 and P3. For the embedded community this is going to slow adoption. What, after all, would be the point in moving to P2 if binary compatibility breaks, likewise for P3? - Given the upgrade cadence available in the embedded space this doesn’t make sense.  - Whether it was advertised or not, the phrase “once bitten twice shy” applies here. 

To summarize, compared to the cloud / hosted world, the cost of adoption of an interim release is a lot higher. The result is likely to be slow to no adoption of P2, or P3. For the BCA this may defeat the point of the preview, as there will be little to no feedback on the production suitability of P2 or P3 from the embedded community.

If the BCA would like that feedback, it would need to commit to providing transition support upfront. This cost might need to be amortized somehow. 

It may not have been intended but there is an existing ecosystem around embedded P1 – millions of devices! As the operating principles pointed to early suggest – *“we empower existing ecosystems instead of disrupting them”* - I’m trying hard to suggest steps that work within the cost and practicality constraints of that ecosystem, so that we don’t disrupt it.

Reviewer 2: this is important:

> A key reason why I personally am excited about the formation of this SIG is that I wish it had existed much earlier, and thus would've been able to support the P1->P2 transition from the get-go.

I could not agree more. Let's keep moving through this, as it's part of the work we wish we'd all done earlier. Alas, each of us is infinitely time-sliced.....

Reviewer 9: I want to also emphasize again that I appreciate that we all want the same thing here at the end of the day! The way I see it we're down to the finer points of how to achieve it, instead of working through fundamental disagreements. And in particular, your feedback on approaches to transitions between WASI versions is well taken, @woodsmc, and I'm very glad we're laying the groundwork here for improving on them going forward!

Reviewer 13: I would like to point out why I think the component model is even more suitable for embedded devices than preview1. WASIp1 is a monolithic POSIX inspired standard, while WASIp2 is highly modular. Just the new timer/clock part is already much more suitable for smallest devices. Whether a wasi-virt polyfill for the filesystem composed with a stock libc or a non-filesystem including reduced libc compilation target is the more efficient solution is an interesting future research topic.

I assume that component fusing technology (after migrating them to a shared everything environment) can reduce the footprint for small embedded devices a lot.

I have investigated into compiling _unmodified components_ with a native compiler and using WIT interfaces without wasm. This can act as a drop-in replacement for the same module compiled to wasm, then thoroughly tested and afterwards compiled to native using wasm2c or w2c2 or wasker or …. If you would like to look at a code example, see https://github.com/cpetig/wit-bindgen/tree/main/crates/cpp/tests/native_strings where I combine Rust and C++ via WIT, all without wasm.

Furthermore I look for a way to map highly efficient zero-copy frameworks like iceoryx(2) into a WIT environment, perhaps GC would provide a good fit, perhaps we need to add a `borrow<list<u8>>` returning resource method.

The idea in all of this is to support polyglot apps, to keep full source code compatibility with the component model ecosystem and the additional option to AoT compile third-party wasm components into the exact same environment, while remaining close to native development.

Reviewer 13: Oh, by the way, I think the lack of SDKs directly targeting preview2 without a component adapter is additionally slowing down p1 to p2 migration for embedded. Extracting core modules from components can easily be done using jco, if you still need some of your p1 tooling :wink:

Reviewer 9: I want to follow up again here because some of my own comments don't sit quite right with me after letting all of this percolate more.

> Not asking for this would be a direct violation of the TSC's mandate, because it'd immediately run into conflict with our mission and operating principles.

This in particular I want to clarify: I fear it might easily be taken as me expressing fear that eSIG itself might violate the BA's principles and act against its mission. That is categorically not what I'm trying to say here; I was trying to explain the TSC's structural position in how the BA works, not indicate that I think violations are bound to happen and will need to be handled. My apologies for not being more clear on this very important point.

Another thing I have reflected on more and that I think is in need of clarification: from the TSC side we've focused on the Component Model, but really I think that's not the right focus, exactly. I think what we really want to ensure is that the in eSIG always be focused on standards-based approaches. Our hope is that the component model will go a very long way—perhaps all of the way—towards covering use cases pursued in eSIG. But yes, that might not be how things turn out, and other avenues might have to be pursued.

In case other approaches have to be pursued, I do think we should ensure strong communication between eSIG and the TSC when it comes to standardization approaches, venues to do this standardization in, etc. The way I interpret the TSC's mandate, that kind of decision should be made at this level to ensure a cohesive approach for the BA overall. I hope that that isn't all that controversial; it _is_ something that hadn't fully crystalized for myself before.

And yet another thing to clarify: I don't think we have a full handle on all the ways the component model can be expressed. Above, @sunfishcode showed a way to target the canonical ABI, no binary components in sight. I think there's lots of potential for making such an approach very smooth. That could even involve making it part of a deployment pipeline instead of the authoring toolchain: that way everything up to deployment is fully standardized components, interoperable with other environments. It'd have the added benefit of retaining more information right up to the point where it's definitely not needed anymore.

(A downside is that it's limited to certain configurations of the canonical ABI, including lowerings to WebAssembly GC, but that's of course part-and-parcel in the embedded space. As discussed above, nobody expects every runtime targeting that space to be able to support the full generality of the component model, or even core WebAssembly.)

I hope these clarifications help make the TSC's position (or in some cases at least mine) clearer!

Reviewer 9: @cpetig thank you for sharing your explorations and findings ❤️ I've long been interested in approaches like the ones you describe, including use of WIT interfaces in native code.

And I agree that WASI 0.2, and WIT based interfaces more generally, should be a great fit for the embedded space in various ways, and look forward to how things unfold on that point.

Reviewer: @woodsmc

> WIT could be used to generate WITx and this would generate preview1 bindings.

I think we are in alignment on this point: I consider that described stack to be an implementation detail of using WIT-based (and therefore component model-based) APIs.

So I would be happy with the following wording, if we want to explicitly call out that approach as allowed:

> New APIs must be based on the component model and its canonical ABI. These APIs may, as an implementation detail, also be rendered in ABI-compatible WITx that is suitable for p1 toolchains.

Author: Thank you to everyone that's contributed to the discussion so far. It's been lively, and I know a lot of energy has spent here. Thank you! - We've made a lot of progress, but there is still more to do. The discussion has seemed overlapping in places and I can see a number of discussions occurring at the same time:

1.	The text review of the proposed charter (the original purpose of the PR)
2.	An alignment discussion between expectations from the BCA and the needs of the embedded community
3.	General excitement for the SIG and some positive input on the prospects for the component model in the embedded space
4.	A request for help and long term support in future standard migrations
 
It’s hard / impossible to do all of this in a PR discussion. It really isn't designed for this. I'd like to take some time to catchup with the embedded folks that commented here, and I'd also like to take some time to explain some of the needs of the embedded ecosystem to the CG; after all, this is community feedback.  Explaining the needs of the embedded ecosystem and the issues / obstacles we currently see will go some way, I hope, to addressing the overlapping discussions we see here.

> Any work the group does which involves standardized extensions to runtimes and their interfaces will enter the appropriate standardization process (e.g. WASM CG, WASI).

Reviewer: ```suggestion
Any work the group does which involves changes to runtimes and their interfaces that are semantically visible to Wasm guests will enter the appropriate standardization process (e.g. the WebAssembly Community Group or its WASI subgroup). Major architectural changes to runtime internals (those not semantically visible to Wasm guests) will follow [the Bytecode Alliance RFC process](https://github.com/bytecodealliance/rfcs/blob/main/accepted/rfc-process.md).
```

> To ensure that all interested members of the WebAssembly community can participate the SIG will arrange regular meetings which rotate between time slots most suitable for Asia, Europe and the Americas. It would therefore be desirable to have three co-chairs, preferably, one from each time-zone who would help facilitate the regular meetings.

Reviewer: Love this! Really appreciate the forethought about including everyone and making sure all time zones can participate.

Author: Thank you!

## Deliverables

> 1. In order to provide scope and guide discussions the SIG will help provide a definition of the lowest addressable device (low footprint device) that WASM and WASI could operate on.

Reviewer: ```suggestion
1. In order to scope and guide discussions, the SIG will define the lowest addressable device (low footprint device) that Wasm and WASI can target.
```

Author: Thanks Nick, this looks cleaner and achieves the same ends.

> 2. The group may provide notes and guides to the embedded community on how to address the specific constraints the ecosystem.

Reviewer: ```suggestion
2. The group may provide notes and guides to the embedded community on how to address the specific constraints of the ecosystem.
```

Author: Thanks for the correction!

> 3. The group may provide notes to the WASM CG, WASI CG on industrial uses cases and technical constraints.

Reviewer: ```suggestion
3. The group may provide notes to the WebAssembly Community Group and its WASI subgroup on industrial uses cases and technical constraints.
```

Author: Cleaner. I like it.

> 4. The group may provide notes and where appropriate technical proof of concepts to the WASM CG and WASI CG on ways to address the specific constraints faced by the Embedded and industrial use cases.

Reviewer: ```suggestion
4. The group may provide notes and where appropriate technical proof of concepts to the WebAssembly Community Group and its WASI subgroup on ways to address the specific constraints faced by the Embedded and industrial use cases.
```

> 5. The group may provide software to support embedded use cases including proof of concept implementations, producer tools, or proposed solutions, which may be specialized for specific use cases or platforms.

Reviewer: ```suggestion
5. The group may propose new Bytecode Alliance hosted projects to support embedded use cases including proof of concept implementations, producer tools, or proposed solutions, which may be specialized for specific use cases or platforms.
```

Author: This would prevent contributions to existing software projects, some of which are already hosted by the bytecode alliance. As it only allows for "new" hosted projects. What if the SIG needed to contribute to an existing project? The WASI SDK, WAMR, or Wasmtime?

Is it really necessary to propose an entire project, in order to show a proof of concept solution to a technical constraint? - See point 4?. This wording would prevent the SIG from doing this.

Is constraint appears quite limiting. Is there a specific concern the TSC has? - Perhaps you could elaborate, and we could address it together.

Reviewer 2: key word here, `may`. Does not say `will`, so I think its fine. I would also think we could add supportive wording: `may propose new Bytecode Alliance hosted projects and may contribute features to existing Bytecode Alliance projects that support....` 

????

Reviewer 2: My guess is that a LOT of wasi-virt things might come out of this, for example, let alone wasi proposals (wasi:usb/i2c,etc), which, by the way, I would assume qualifies us for "projects" in any case. Otherwise, we'll end up writing guidance for how wasm and component usage seems to have tradeoff event horizons for various use cases, and that's critical for new entrants into the ecosystem to think clearly more rapidly than those of us muddling our way through at the moment..... 

What do you think? @fitzgen

Reviewer 2: oooo and wasi-libc work.... frankly, I see this as DEFINITELY producing BCA project contributions at a minimum. the proposals that do not yet exist will be new, from my pov.

Reviewer 9: oh for sure: we definitely don't want to forbid contributions to existing projects! The suggested edit was more about making explicit that *new* projects might come out of this work.

Reviewer: Yeah the intention was definitely not to prohibit contributing to existing projects; just about the process for creating new projects.

For clarity, it might be easiest to split this into two bullet points: one for existing and one for new projects.

Author: Folks, this all sounds good. Great suggestion from @squillace and @fitzgen to clean this up with two bullet points:

> 5. The group may contribute features to existing Bytecode Alliance projects that support embedded / industrial use cases including proof of concept implementations, producer tools or proposed solutions.
> 6. The group may propose new Bytecode Alliance hosted projects to support embedded use cases including proof of concept implementations, producer tools, or proposed solutions, which may be specialized for specific use cases or platforms.
